### ISSUE TITLE
Few tests are sometimes in failure

### ISSUE THREAD

**Issue by class101** [NationalSecurityAgency/ghidra#2854]
Hello everyone,

I'm new to the project and setting up the developer environment, everything goes well and works perfectly but I have one concern.

While executing unit and integration tests, I'm getting 69 test failures on 12226 total tests

What I note is that most of the failures obtained come from running the tests from the command line with `gradle` command, but when the same test is run in Eclipse, it returns Success. I will post my system specifications at the end of this ticket.

### Bash output test

The following test for example works fine executed in Eclipse, but under gradle command the output is slightly different
https://github.com/NationalSecurityAgency/ghidra/blob/b69c3d6c50c666ccece1fcc25e5ab7561ba57072/Ghidra/Debug/Debugger-agent-gdb/src/test/java/agent/gdb/ffi/linux/PtyTest.java#L130

The expected output (captured under Eclipse)
```bash
New Pty: /dev/pts/3 at (51,53)
log: echo test
log: BASH:echo test
log: test
log: BASH:exit 3
```

The output captured with `gradle parallelCombinedTestReport` or `gradle combinedTestReport` with or without Xvfb headless
There are some unexpected extra symbols `?[?2004h` and `?[?2004l` that I'm trying to understand from where they come from.

```
New Pty: /dev/pts/0 at (95,96)
New Pty: /dev/pts/0 at (95,96)
New Pty: /dev/pts/0 at (95,96)
log: echo test
log: ?[?2004hBASH:echo test
log: ?[?2004l
log: test
log: ?[?2004hBASH:cat
New Pty: /dev/pts/0 at (95,96)
New Pty: /dev/pts/0 at (95,96)
log: echo test
log: ?[?2004hBASH:echo test
log: ?[?2004l
log: test
log: ?[?2004hBASH:exit 3
New Pty: /dev/pts/0 at (95,96)
New Pty: /dev/pts/0 at (95,96)

```

### Debugger Breakpoint Marker Plugin

Another failure reported in the following test, it works as expected executed in Eclipse, but with gradle command I get the following result

https://github.com/NationalSecurityAgency/ghidra/blob/b69c3d6c50c666ccece1fcc25e5ab7561ba57072/Ghidra/Debug/Debugger/src/test/java/ghidra/app/plugin/core/debug/gui/breakpoint/DebuggerBreakpointMarkerPluginTest.java#L312

```java
junit.framework.AssertionFailedError: expected:<[Disable Breakpoint, Set Breakpoint, Toggle Breakpoint, Clear Breakpoint]> but was:<[]>
  at generic.test.AbstractGTest.waitForCondition(AbstractGTest.java:477)
  at generic.test.AbstractGTest.waitForCondition(AbstractGTest.java:433)
  at ghidra.app.plugin.core.debug.gui.AbstractGhidraHeadedDebuggerGUITest.waitForPass(AbstractGhidraHeadedDebuggerGUITest.java:239)
  at ghidra.app.plugin.core.debug.gui.AbstractGhidraHeadedDebuggerGUITest.assertMenu(AbstractGhidraHeadedDebuggerGUITest.java:280)
  at ghidra.app.plugin.core.debug.gui.breakpoint.DebuggerBreakpointMarkerPluginTest.testProgramBreakpointPopupMenus(DebuggerBreakpointMarkerPluginTest.java:320)
  at generic.test.TestThread.run(TestThread.java:106)
```
I'm running the test headless in CI environment with 8 forks, and using Xvfb and DISPLAY=:99 in Eclipse did not help to reproduce the errors in eclipse, it just works as expected so I don't think it has something to do with this.

System
```
OS: Arch Linux x86_64 
Kernel: 5.11.7-zen1-1-zen 
Uptime: 2 hours, 5 mins 
Packages: 1276 (pacman) 
Shell: bash 5.1.4 
Resolution: 1920x1080 
DE: Cinnamon 4.8.6 
WM: Mutter (Muffin) 
WM Theme: Mojave-dark (Mojave-dark) 
Theme: Mojave-dark [GTK2/3] 
Icons: McMojave-circle [GTK2/3] 
Terminal: guake 
CPU: Intel i7-4770K (8) @ 3.900GHz 
GPU: NVIDIA GeForce GTX 980 
GPU: NVIDIA GeForce GTX 980 
Memory: 4981MiB / 15937MiB
```

**Comment by dragonmacher** [comment]
I believe we still have some timing sensitivities in our corpus of tests.     I have personally seen that each new system (such as changing OS and/or hardware) we run tests on will expose a group of test failures that were not failing on pre-existing test machines.    We have been working for years to rid the system of these test timing sensitivities as we find them.

In this case, 69 failures out of 12k+ tests fits the pattern that we have seen.   Unfortunately, without having your exact test setup ro reproduce the issue, it would be quite challenging for us to fix these types of failures.   

> I'm running the test headless in CI environment with 8 forks, and using Xvfb and DISPLAY=:99 in Eclipse did not help to reproduce the errors in eclipse, it just works as expected so I don't think it has something to do with this,

This setup should work just fine.

> How does your tests performs on your CI server with 40 CPUs, always success without false positive as such ? or are the false positive results common noises to ignore ?

As I mentioned above, when we get these issues in our environment, we work to fix them.   For your case, it may be easiest for you to keep a list of these known offending tests so you know that you can ignore them as you  are working in the codebase.

**Comment by class101** [comment]
Yeah I will eventually ignore them as known noise, was much to know if you already heard of such possible problems while setting up a new env.

@dragonmacher Out of curiosity, how long does it takes on your environment to execute all the unit and integration tests ?

**Comment by class101** [comment]
For the bash output it seem caused by colors in the bash output and `TERM`

With `TERM=xterm` or `TERM=xterm-256color` the bash tests fails

With `TERM=` it is success

**Comment by class101** [comment]
Another note about build times and a few unexpected test failures, I pin point this the use of the following jvm option in `test.gradle` and `javaTestProject.gradle`

### `-Djava.compiler=NONE`

As a quick comparison, below are a few tests runs I did and the time it took to complete
 
- unit tests + `-Djava.compiler=NONE` = 2h20min
- unit tests + `--parallel` + `-Djava.compiler=NONE` = 1h40min
- unit tests + `--parallel` + java.compiler removed = 10min (and 34 unexpected failures, success now)

To diagnose I took the following test who gave a unpexected long time to execute and in the end, an unexpected NullPointerException.

https://github.com/NationalSecurityAgency/ghidra/blob/b69c3d6c50c666ccece1fcc25e5ab7561ba57072/Ghidra/Framework/SoftwareModeling/src/test/java/ghidra/app/plugin/assembler/sleigh/AARCH64BEAssemblyTest.java#L69

I realized it took 36sec. to execute in Eclipse (fast), but on command line

```
DISPLAY=:99 gradle SoftwareModeling:test --tests AARCH64BEAssemblyTest.testAssemble_and_w0_w0_0xfffffffe
```
It took like 2min or more that the first log line entry `Expected tree: ` was not yet even printed ! It is almost immediate under Eclipse.

So I took all the jvmArgs you used to set in `javaTestProject.gradle` until I discovered that with `-Djava.compiler=NONE` under Eclipse, it was slow and I was able to reproduce the exact problem noted on command line.

Here, removing `-Djava.compiler=NONE` radically increased tests execution speeds from more than a hour to 10min, and it is night and day it performs better without on my system, logs are fluid, java looks like fast, with it, java is very very busy to perform your tests.

**Comment by dragonmacher** [comment]
> @dragonmacher Out of curiosity, how long does it takes on your environment to execute all the unit and integration tests ?

Typically, for a headless server env we see total test times running around 2-.2.5 hours.

>unit tests + -Djava.compiler=NONE = 2h20min
unit tests + --parallel + -Djava.compiler=NONE = 1h40min
unit tests + --parallel + java.compiler removed = 10min (and 34 unexpected failures, success now)

Keep in mind that the `parallel` does not run all tests.  There is a subset of tests that does not fit well into the parallel format.   The slower and more sensitive tests will get run during a full non-parallel test session.     The parallel option is meant more for developers to quickly run most of the tests during development, with the full set of tests being used during the CI process.

> -Djava.compiler=NONE

This option is interesting to me.  I will have to do some research to see why this option was introduced.   I have a feeling it has been there for a long time.   It is probably not needed for our current state of testing.

**Comment by dragonmacher** [comment]
> Ideally you should support parallelism, the spirit of the gradle today, and write correct thread safe tests.

Writing thread safe code, and correct code, in general is surprisingly difficult, even more so for testing due to the inherent asynchronous nature of testing.   Very few projects have the resources to retrofit old code, especially when it comes to rewriting tests, unfortunately.    In the interest of better CI and more responsive feedback,  we started down the path of making the tests parallel, which required quite a bit of test fixing along the way.    As you have seen, the test framework is currently somewhere between perfect and less than ideal.

**Comment by class101** [comment]
From what I have tested, the tests are very well done and I think most of your developers has parallelism in mind because out of all the tests, only the tests from the debugger branch are causing problem, the following ones, yet I just had to fix the bash TERM variable, and another case where an abstract was initializing the application without verifying if it was already.

```
DebuggerBreakpointMarkerPluginTest. testProgramBreakpointPopupMenus
DebuggerBreakpointMarkerPluginTest. testProgramNoBreakPopupMenus
DebuggerBreakpointMarkerPluginTest. testTraceBreakpointPopupMenus
DebuggerBreakpointMarkerPluginTest. testTraceNoBreakPopupMenus
DebuggerBreakpointsProviderTest. testActionClearSelectedBreakpoints
DebuggerBreakpointsProviderTest. testActionEnableSelectedBreakpoints
DebuggerBreakpointsProviderTest. testPopupActionsOnBreakpointSelections
DebuggerListingProviderTest. testAutoReadMemoryReads
DebuggerListingProviderTest. testMemoryStateBackgroundColors
DebuggerListingProviderTest. testSyncToStaticListingMarksMappedTrackedRegister
DebuggerModulesProviderTest. testActionCaptureTypes
DebuggerModulesProviderTest. testActionMapSections
DebuggerModulesProviderTest. testPopupActionsOnModuleSelections
DebuggerModulesProviderTest. testPopupActionsOnSectionSelections
DebuggerRegistersProviderTest. testActionCreateSnapshot
DebuggerTargetsProviderTest. testModelActivationOnClick
DebuggerLogicalBreakpointServiceTest. testRecordTraceThenAddBreakpointThenOpenTrace
DebuggerTraceManagerServiceTest. testSynchronizeFocusTraceToModel
```

**Comment by nsadeveloper789** [comment]
Coming to this a bit late.... Thanks for the pointer regarding TERM. That had been vexing me for some time. I'm unable to examine that list of failing tests ATM, as we're still quite busy with other parts of the debugger.

**Comment by nsadeveloper789** [comment]
Do you have the command line to produce that list handy? I'd like to check it against my current working copy.

**Comment by class101** [comment]
You can quickly reproduce this with 

`DISPLAY=:99 gradle --parallel Debugger:test`

Basically I think phase name is always &lt;Project Name&gt;:test and also you can test specific &lt;Classes&gt;.&lt;methodName&gt; if you add 

`--tests DebuggerBreakpointMarkerPluginTest.testProgramBreakpointPopupMenus`

Can be repeated to test different classes and also wildcards works, more example [here](https://docs.gradle.org/current/userguide/java_testing.html#simple_name_pattern)

Also if you are going to test parallel builds, you have to remove forkEvery 1 in javaTestProject or set it back to 0 the default

https://github.com/NationalSecurityAgency/ghidra/blob/cbe83dcd7853dc09e7c05113a4cf9a5fb0fb1fad/gradle/javaTestProject.gradle#L23

**Comment by class101** [comment]
I have found that after adding `-noverify` and `-XX:TieredStopAtLevel=1` (based on the following [article](https://phauer.com/2017/increase-jvm-development-productivity/)), it improves the execution of unit tests from 10min to 5min here :)

I did not commit yet all the options to the fork but to make this simpler to test, I'm attaching a patch and after applying all you have to do is `DISPLAY=:99 gradle --parallel unitTestReport`

I have added a few options like

- -PfailFast : Stop the tests as soon a test failure happens (default is off)
- -PjvmForks : Control the num of workers of a gradle task (default is 8)
- -PjvmDebug : Enable debug mode (default is off)
- -PforkEvery=8 : The maximum number of test classes to execute in a forked test process. (default is 0)
- 
[0001-Unit-test-execution-down-to-5min.patch.txt](https://github.com/NationalSecurityAgency/ghidra/files/6197017/0001-Unit-test-execution-down-to-5min.patch.txt)

**Comment by dragonmacher** [comment]
We are exploring using some or all of the changes you have presented.    What is becoming clear though, after digesting this information is that we likely have multiple test use cases at play for Ghidra.

The test modes we currently support in some form:
- Full Tests (Synchronous) - we run all tests at least once a day, non-parallel, regardless of code changes
- Most Tests (Parallel) - on-demand we support the build system running the bulk of our tests (roughly 10k tests) in parallel
- Targeted Tests - run tests when a branch is pushed, attempting to run only tests related to changes in the branch  (this last mode does not work very well, due mostly to the fact that we have poor tracing from non-test code to the relevant test(s)).

From what I can tell, the way you are changing the tests above is more typical to a standard development workflow, where the developer can run tests from gradle as they develop.       This is not a mode that we use frequently in our development process, which partially explains why the tests run so slowly when manually executing them using gradle.     We rely on test servers to allow users to execute the suite of tests remotely.  We do this for a few reasons, one of which being that developers would like to be able to continue working while tests are running without having them steal CPU cycles or steal window focus while they are working.

I think there is value in being able to run the test in the fashion, especially if they can be run relatively quickly.

**Comment by class101** [comment]
No problem, not asking to change the build or things in place, they are much comments from my point of view, I'm not always right, often wrong, but at least maybe my remarks will improve the quality of life of a remote developer.

I'm trying to understand how it works when you don't have the continous integration available. Because like you, I used to work in big organizations, I used to setup a lot of Maven build and Jenkins all around, so I know when things are working and in place, you really don't want to touch it hehe :)

I'm much trying to understand how it works as remote random developer as I don't have a lot of experience is participating in open source projects on the web, the only way to remotely check if I'm bringing regression are tests I execute here right ?

**Comment by dragonmacher** [comment]
>  so I know when things are working and in place, you really don't want to touch it hehe, 

This

>  the only way to remotely check if I'm bringing regression are tests I execute here right ? So if yes, 2h+ to get the result is a bit restricting

This is a good point and is a use case that we really have not been supporting well.   We put a lot of emphasis on iterative development and refactoring.   As we constantly improve this area, this information will help us as we change the test code to better understand how the test system is being used.

**Comment by class101** [comment]
thank you for the help and all your time anyway, this is very appreciated. I have updated my comment https://github.com/NationalSecurityAgency/ghidra/issues/2854#issuecomment-805773229 

To include another tip to detect when --parallel is used `gradle.properties.startParameter.parallelProjectExecutionEnabled` I think I will do that to manage exclusions for the moment, why not to find a way to `forkEvery` but only the problematic tests

**Comment by class101** [comment]
To note I did few more tests today, the use of `--parallel` is not really a big improvment, it is a win of 3min, and without, it takes 9min43s exactly with the following options matching what you have in files actually

`DISPLAY=:99 gradle unitTestReport -PjvmForks=1 -PforkEvery=1`

I think if you are concerned about improving your build time, whether it's on a developer workstation or in continuous integration you will have to consider in priority

- removal of -Djava.compiler=NONE
- use of `-noverify` and `-XX:TieredStopAtLevel=1`
- debug mode by default to off
